src/crawlee/_utils/persistent_object.py

```diff
@@ -0,0 +1,152 @@
+
+        This method must be called before using the persistent object. It loads the saved state
+        if persistence is enabled and registers the object to listen for PERSIST_STATE events.
+
+        Returns:
+            The loaded state model
+        """
+        if not self._persistence_enabled:
+            self._state = self._default_state.model_copy(deep=True)
+            return self.get()
+
+        self._key_value_store = await KeyValueStore.open(
+            name=self._persist_state_kvs_name, id=self._persist_state_kvs_id
+        )
+
+        await self._load_saved_state()
+
+        event_manager = service_locator.get_event_manager()
+        event_manager.on(event=Event.PERSIST_STATE, listener=self.persist_state)
+
+        return self.get()
+
+    async def teardown(self) -> None:
+        """Clean up resources used by the persistent object.
+
+        If persistence is enabled, this method deregisters the object from PERSIST_STATE events
+        and persists the current state one last time.
+        """
+        if not self._persistence_enabled:
+            return
+
+        event_manager = service_locator.get_event_manager()
+        event_manager.off(event=Event.PERSIST_STATE, listener=self.persist_state)
+        await self.persist_state()
+
+    def get(self) -> TStateModel:
+        """Get the current state."""
+        if self._state is None:
+            raise RuntimeError('Persistent state has not yet been loaded')
+
+        return self._state
+
+    async def reset(self) -> None:
+        """Reset the state to the default values and clear any persisted state.
+
+        Resets the current state to the default state and, if persistence is enabled,
+        clears the persisted state from the KeyValueStore.
+        """
+        self._state = self._default_state.model_copy(deep=True)
+
+        if self._persistence_enabled:
+            if self._key_value_store is None:
+                raise RuntimeError('Persistent object has not yet been initialized')
+
+            await self._key_value_store.set_value(self._persist_state_key, None)
+
+    async def persist_state(self, event_data: EventPersistStateData | None = None) -> None:
+        """Persist the current state to the KeyValueStore.
+
+        This method is typically called in response to a PERSIST_STATE event, but can also be called
+        directly when needed.
+
+        Args:
+            event_data: Optional data associated with a PERSIST_STATE event
+        """
+        self._log.debug(f'Persisting state of the Statistics (event_data={event_data}).')
+
+        if self._key_value_store is None or self._state is None:
+            raise RuntimeError('Persistent object has not yet been initialized')
+
+        if self._persistence_enabled:
+            await self._key_value_store.set_value(
+                self._persist_state_key,
+                self._state.model_dump(mode='json', by_alias=True),
+                'application/json',
+            )
```

Comment:
Maybe we can use the [`ensure_context`](https://github.com/apify/crawlee-python/blob/v0.6.8/src/crawlee/_utils/context.py#L10) decorator?

Reply:
It feels a bit weird because most of the components that use this will also have `ensure_context`, but say the word and I'll do it.

Reply:
Yep, because all of them follow the same pattern, I don't see an issue with it, now there's just a different check mechanism and different error message.